File: app/src/main/java/org/scp/gymlog/util/FormatUtils.java
package org.scp.gymlog.util;

import android.util.DisplayMetrics;
import android.util.TypedValue;

import java.math.BigDecimal;
import java.text.DecimalFormat;
import java.text.Format;
import java.text.ParseException;
import java.util.Arrays;
import java.util.Locale;

public class FormatUtils {

    private static final Format FORMAT;
    static {
        DecimalFormat decimalFormat = (DecimalFormat) DecimalFormat.getInstance(Locale.getDefault());
        decimalFormat.setParseBigDecimal(true);
        decimalFormat.setGroupingUsed(false);
        FORMAT = decimalFormat;
    }

    public static BigDecimal toBigDecimal(String string) {
        if (string.isEmpty())
            return BigDecimal.ZERO;

        try {
            return (BigDecimal) FORMAT.parseObject(string);

        } catch (NumberFormatException | ParseException e) {
            e.printStackTrace();
            return BigDecimal.ZERO;
        }
    }

    public static int toInt(String string) {
        try {
            return Integer.parseInt(string);
        } catch (NumberFormatException e) {
            e.printStackTrace();
            return 0;
        }
    }

    public static String toString(BigDecimal bigDecimal) {
        return FORMAT.format(bigDecimal);
    }

    public static int toDp(DisplayMetrics displayMetrics, int value) {
        return (int) toDpFloat(displayMetrics, value);
    }

    public static float toDpFloat(DisplayMetrics displayMetrics, int value) {
        return TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP,
                value,
                displayMetrics
        );
    }

    @SafeVarargs
    public static <T> boolean isAnyOf(T needle, T... haystack) {
        return Arrays.asList(haystack).contains(needle);
    }
}

File: app/src/main/java/org/scp/gymlog/util/FormatUtils.kt
package org.scp.gymlog.util

import android.util.DisplayMetrics
import android.util.TypedValue
import android.widget.TextView
import java.math.BigDecimal
import java.text.DecimalFormat
import java.text.Format
import java.text.ParseException
import java.util.*

object FormatUtils {

    private val FORMAT: Format
    init {
        val decimalFormat = DecimalFormat.getInstance(Locale.getDefault()) as DecimalFormat
        decimalFormat.isParseBigDecimal = true
        decimalFormat.isGroupingUsed = false
        FORMAT = decimalFormat
    }

    fun String.safeBigDecimal(): BigDecimal {
        return if (this.isBlank())
            BigDecimal.ZERO
        else {
            try {
                FORMAT.parseObject(this) as BigDecimal
            } catch (e: NumberFormatException) {
                e.printStackTrace()
                BigDecimal.ZERO
            } catch (e: ParseException) {
                e.printStackTrace()
                BigDecimal.ZERO
            }
        }
    }

    fun String.safeInt(): Int {
        return try {
            this.toInt()
        } catch (e: NumberFormatException) {
            0
        }
    }

    fun BigDecimal.toLocaleString(): String {
        return FORMAT.format(this);
    }

    var TextView.bigDecimal: BigDecimal
        get() = this.text.toString().safeBigDecimal()
        set(value) { text = value.toLocaleString() }

    var TextView.integer: Int
        get() = this.text.toString().safeInt()
        set(value) { text = value.toString() }

    fun toDp(displayMetrics: DisplayMetrics, value: Int): Int {
        return toDpFloat(displayMetrics, value).toInt()
    }

    fun toDpFloat(displayMetrics: DisplayMetrics, value: Int): Float {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            value.toFloat(),
            displayMetrics
        )
    }

    fun <T> T.isAnyOf(vararg haystack: T): Boolean {
        return listOf(*haystack).contains(this)
    }
}